File: SITA23/Complexité/exos.py
def sum_m(lst):
    max_i = 0
    max_j = 0
    max_sum = 0

    for i in range(len(lst)):
        current_sum = 0
        for j in range(i, len(lst)):
            current_sum += lst[j]
            if current_sum > max_sum:
                max_sum = current_sum
                max_i = i
                max_j = j

    return max_sum, max_i, max_j

if __name__ == "__main__":
    lst = [1, -2, 1, 2, -1, 3, -2, 1]
    max_sum, max_i, max_j = sum_m(lst)
    print(f"Max sum: {max_sum}, from index {max_i} to {max_j}")

File: SITA23/Complexité/exos.c
#include <stdio.h>

void sum_m(int list[], int length, int *max_sum, int *max_i, int *max_j) {
    *max_i = 0;
    *max_j = 0;
    *max_sum = 0;

    for (int i = 0; i < length; i++) {
        int sum = 0;
        for (int j = i; j < length; j++) {
            sum += list[j];
            if (sum > *max_sum) {
                *max_sum = sum;
                *max_i = i;
                *max_j = j;
            }
        }
    }
}

int main() {
    int list[] = {1, -2, 1, 2, -1, 3, -2, 1};
    int length = sizeof(list) / sizeof(list[0]);
    int max_sum, max_i, max_j;

    sum_m(list, length, &max_sum, &max_i, &max_j);

    printf("Max sum: %d, from index %d to %d\n", max_sum, max_i, max_j);

    return 0;
}